File: src/learning_gem5/part2/PowerModelFunc.py
from m5.SimObject import SimObject
from m5.params import *
from m5.objects.PowerModelState import PowerModelState

# Dynamic and static power equations represented by arithmetic operators than strings in MathExprPowerModel
class PowerModelFunc(PowerModelState):
    type = "PowerModelFunc"
    cxx_header = "learning_gem5/part2/power_model_func.hh"
    cxx_class = "gem5::PowerModelFunc"

    # Equations for dynamic and static power in Watts
    # Equations may use gem5 stats ie. "1.1*ipc + 2.3*l2_cache.overall_misses"
    # It is possible to use automatic variables such as "temp"
    # You may also use stat names (relative path to the simobject)
    dyn = Param.PyFunc("Function to call for Dynamic Power")
    stat = Param.PyFunc("Function to call for Static Power")

File: src/learning_gem5/part2/power_model_func.cc
#include "learning_gem5/part2/power_model_func.hh"

#include "base/trace.hh"
#include "debug/PowerModelFunc.hh"

namespace gem5{

PowerModelFunc::PowerModelFunc(const Params &p)
           :  PowerModelState(p), dyn(p.dyn), st(p.st)
        {
           // Bind PyFunc parameters into functions to be called in this SimObj

           dyn_func =
                   pybind11::reinterpret_borrow<pybind11::function>(dyn);
           st_func =
                   pybind11::reinterpret_borrow<pybind11::function>(st);
        }
/*
void
PowerModelFunc::startup()
{

}
*/
} // namespace gem5
